plugin/src/main/java/org/jenkinsci/plugins/structs/SymbolLookup.java

```diff
@@ -31,12 +34,39 @@
 public class SymbolLookup {
     private final ConcurrentMap<Key,Object> cache = new ConcurrentHashMap<Key, Object>();
 
+    private final ConcurrentMap<Key,Object> noHitCache = new ConcurrentHashMap<Key, Object>();
+
+    static final Object NO_HIT = new Object();
+
     @Inject
     PluginManager pluginManager;
 
     @Inject
     Jenkins jenkins;
 
+    Set<String> pluginNames = Collections.EMPTY_SET;
+
+    private static HashSet<String> pluginsToNames(List<PluginWrapper> plugins) {
+        HashSet<String> pluginNames = new HashSet<String>(plugins.size());
+        for (PluginWrapper pw : plugins) {
+            pluginNames.add(pw.getShortName());
+        }
+        return pluginNames;
+    }
+
+    /** Update list of plugins used and purge the noHit cache if plugins have been added
+     */
+    private synchronized void checkPluginsForChangeAndRefresh() {
+        List<PluginWrapper> wrap = pluginManager.getPlugins();
+        Set<String> names = pluginsToNames(wrap);
+
+        if (wrap.size() != pluginNames.size() || !(pluginNames.containsAll(names))) {
+            this.pluginNames = names;
+            noHitCache.clear();
```

Comment:
Deserves a test I think: create a tiny plugin with a symbol, put the `*.hpi` into `src/test/resources/`, then have the test look up that symbol, call `PluginManager.somethingOrOther` to install the plugin, look up the symbol again.